### PR TITLE
.NET MAUI is now a "workload"

### PR DESCRIPTION
**Pull request by jonathanpeppers**
### Description of Change ###

TDLR: Microsoft.Maui.* NuGet packages are removed in favor of .NET
Workloads.

See [`src/Workloads/README.md`](https://github.com/jonathanpeppers/maui/blob/dotnet-workload/src/Workload/README.md) for full details.

The idea, is a project to be able to set `$(UseMaui)`:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0-android</TargetFramework>
    <OutputType>Exe</OutputType>
    <UseMaui>true</UseMaui>
  </PropertyGroup>
</Project>
```

`$(UseMaui)` automatically brings in the following workload packs:

* `Microsoft.NET.Sdk.Maui`
* `Microsoft.Maui.Controls.Sdk`
* `Microsoft.Maui.Resizetizer.Sdk`
* `Microsoft.Maui.Core.Ref.[platform]`
* `Microsoft.Maui.Core.Runtime.[platform]`
* `Microsoft.Maui.Controls.Ref.[platform]`
* `Microsoft.Maui.Controls.Runtime.[platform]`
* `Microsoft.Maui.Dependencies`
* `Microsoft.Maui.Essentials.Ref.[platform]`
* `Microsoft.Maui.Essentials.Runtime.[platform]`
* `Microsoft.Maui.Extensions`
* `Microsoft.Maui.Templates`

BlazorWebView is an addition to MAUI, project can currently opt into
it by adding `.Razor` to the `Sdk` attribute.

`<Project Sdk="Microsoft.NET.Sdk.Razor">` sets
`$(UsingMicrosoftNETSdkRazor)`, which triggers the MAUI workload to
include:

* `Microsoft.AspNetCore.Components.WebView.Maui`

This will automatically add these dependencies:

```xml
<PackageReference Include="Microsoft.AspNetCore.Authorization" />
<PackageReference Include="Microsoft.AspNetCore.Components.WebView" />
<PackageReference Include="Microsoft.AspNetCore.Components.WebView.Maui" />
<PackageReference Include="Microsoft.JSInterop" />
```

If you are a .NET 6 project, but don't want to use
Microsoft.Maui.Controls you could bring in partial parts of MAUI.

`$(UseMauiAssets)` brings in `Microsoft.Maui.Resizetizer.Sdk`.

`$(UseMauiCore)` brings in:

* `Microsoft.Maui.Core.Ref.[platform]`
* `Microsoft.Maui.Core.Runtime.[platform]`

`$(UseMauiEssentials)` brings in:

* `Microsoft.Maui.Essentials.Ref.[platform]`
* `Microsoft.Maui.Essentials.Runtime.[platform]`

Special files:

* `AutoImport.props` - defines the default includes (or wildcards) for
  Maui projects will go. Note that this is imported by *all* .NET 6
  project types -- *even non-mobile ones*.
* `WorkloadManifest.json` - general .NET workload configuration
* `WorkloadManifest.targets` - imports `Microsoft.Maui.Controls.Sdk` when
  `$(UseMaui)` is `true`. Note that this is imported by *all* .NET 6
  project types -- *even non-mobile ones*.

For further details about .NET Workloads, see these .NET design docs:

* [.NET Optional SDK Workloads](https://github.com/dotnet/designs/blob/main/accepted/2020/workloads/workloads.md)
* [Workload Resolvers](https://github.com/dotnet/designs/blob/main/accepted/2020/workloads/workload-resolvers.md)
* [Workload Manifests](https://github.com/mhutch/designs/blob/b82449a228c0addb95b5a4995bb838749ea6f8cc/accepted/2020/workloads/workload-manifest.md)

### PR Checklist ###

- [ ] Targets the correct branch 
- [ ] Tests are passing (or failures are unrelated)
- [ ] Targets a single property for a single control (or intertwined few properties)
- [ ] Adds the property to the appropriate interface
- [ ] Avoids any changes not essential to the handler property
- [ ] Adds the mapping to the PropertyMapper in the handler
- [ ] Adds the mapping method to the Android, iOS, and Standard aspects of the handler
- [ ] Implements the actual property updates (usually in extension methods in the Platform section of Core)
- [ ] Tags ported renderer methods with [PortHandler]
- [ ] Adds an example of the property to the sample project (MainPage)
- [ ] Adds the property to the stub class
- [ ] Implements basic property tests in DeviceTests

#### Does this PR touch anything that might effect accessibility?

No

### CODE REVIEW

**Review by feartDev — APPROVED** [review]

**Comment by jonathanpeppers** [comment]
Wanted to save this screenshot from https://github.com/dotnet/maui/commit/99c2b8cea29976d99fbe8d8fb38bb3624d6e5b40:

image

I've come back to work on this, and this PR isn't going to be green again for a while. 👀

**Comment by jonathanpeppers** [comment]
https://github.com/dotnet/maui/pull/603/commits/e3bf1a35ef113cb898db9d7f54b56bffd08a67c7 was green, about to rebase & squash some more.

**Comment by jonathanpeppers** [comment]
https://github.com/dotnet/maui/pull/603/commits/9b326febf52768184faef52c297dbd4af4c29441 was green, need to rebase for new XamlG source generator.

**Comment by jonathanpeppers** [comment]
We will need this before the samples run on Android: https://github.com/xamarin/xamarin-android/pull/5965

**Review by PureWeen — APPROVED** [review]
- Ran through our different cake targets (NET6/WINUI) to make sure everything ran
- installed the workloads and created a new maui project from the template to ensure workloads are setup correctly

Be good to still have @mattleibow  and/or @Redth sanity check the workload bits

We can probably do this as a different PR. But it'd be nice to get something that just launches the samples and validates that they started up correctly on  a device

**Comment by mattleibow** [comment]
Should `UseEssentials` be maybe `UseMauiEssentials` - both for consistency as well as the fact that "essentials" is very generic.

File: src/Controls/samples/Controls.Sample.iOS/Maui.Controls.Sample.iOS-net6.csproj [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk.Razor">
```
Comment (rmarinho) [inline src/Controls/samples/Controls.Sample.iOS/Maui.Controls.Sample.iOS-net6.csproj:1]: why this change?

File: src/Controls/src/Build.Tasks/Controls.Build.Tasks-net6.csproj [hunk 2]
```diff
@@ -33,13 +34,6 @@
 		<!-- <PackageReference Include="StrongNamer" Version="0.2.5" PrivateAssets="all" GeneratePathProperty="true" /> -->
 	</ItemGroup>
 
-	<PropertyGroup>
-		<PackageId>Microsoft.Maui.Controls.Build.Tasks</PackageId>
-		<IncludeBuildOutput>False</IncludeBuildOutput>
-		<NoWarn>$(NoWarn);NU5100;NU5128;</NoWarn>
```
Comment (rmarinho) [inline src/Controls/src/Build.Tasks/Controls.Build.Tasks-net6.csproj:39]: why we are removing this?

File: src/Controls/samples/Controls.Sample.iOS/Maui.Controls.Sample.iOS-net6.csproj (hunk 1 above)
Comment (rmarinho) [reply src/Controls/samples/Controls.Sample.iOS/Maui.Controls.Sample.iOS-net6.csproj:1]: oh i understand now..

Comment (jonathanpeppers) [reply src/Controls/samples/Controls.Sample.iOS/Maui.Controls.Sample.iOS-net6.csproj:1]: Yes, the `.Razor` is what brings in BlazorWebView.

I would rather make `UseBlazor=true` bring this in, but I don't see a way to do this from a workload.

A workload doesn't have an `Sdk.props` file, which is what we'd need to import the `Razor` Sdk ourselves.

File: src/Controls/src/Build.Tasks/Controls.Build.Tasks-net6.csproj (hunk 2 above)
Comment (jonathanpeppers) [reply src/Controls/src/Build.Tasks/Controls.Build.Tasks-net6.csproj:39]: This package no longer produces a NuGet package (uses `IsPackable=false`), so we can remove all NuGet-related properties.

**Comment by jonathanpeppers** [comment]
I marked this "ready for review" because we have this Android fix now: ccf0a6b8046da229d01dc285c9fd288d5bb61fb1

I can build/run the sample with:

```
> git clean -dxf .\src\Controls\samples\ 
> .\bin\dotnet\dotnet build .\src\Controls\samples\Controls.Sample.Droid\Maui.Controls.Sample.Droid-net6.csproj -t:Run -p:UseWorkload=true
```
image

**Review by mattleibow — APPROVED** [review]
I think I understand most of what is going on with the MSBuild bits. 😅 

We need to have a look at cleaning up that .nuspec folder as well as hopefully start to get more things in the net6 sln. But that is not for here.

**Comment by jonathanpeppers** [comment]
https://github.com/dotnet/maui/pull/603/commits/1f859a15138e5c9d485f5beb4e391fecdb6d64e8 was "working", just saving a note here before I rebase and make some more changes.

**Comment by jonathanpeppers** [comment]
This is ready to review again, changes since last time:
* BlazorWebView is left unchanged as a NuGet package. We redistribute it in `dotnet/library-packs` with the workload, and changing to `Sdk="Microsoft.NET.Sdk.Razor"` will automatically bring it in your project.
* Microsoft.Extensions dependencies are repackaged in a `Microsoft.Maui.Extensions` workload pack.
* A new `$(MicrosoftMauiSdkVersion)` property can be used to pin a specific version of certain packs. This doesn't work for MSBuild tasks, but it does work for things like Essentials & Controls that are used at runtime in an app.
